### PR TITLE
Make BundleGraph parameterized on a particular Bundle type

### PR DESCRIPTION
**Pull request by wbinnssmith**
This allows the public `BundleGraph` to be constructed with a given Bundle factory, and makes the `BundleGraph` polymorphic on anything that implements `IBundle`, allowing the public `BundleGraph` to return either `Bundle`s or `NamedBundle`s.

The `MutableBundleGraph` only returns `Bundle`s for now, as that is all that is needed.

This allows namers to receive `BundleGraph<Bundle>` — an immutable BundleGraph that produces unnamed `Bundle`s — and things like runtimes and packagers to receive `BundleGraph<NamedBundle>`, a BundleGraph which produces `NamedBundle`s

As a simplification, this removes the caching behavior for re-using public BundleGraph wrappers. These are not created at the rate of other public wrappers, and I believe that the way we pass them to plugins, they don't really benefit from referential equality guarantees. cc @devongovett 

Test Plan: `yarn flow check && yarn test`

### CODE REVIEW

File: packages/core/core/src/public/BundleGraph.js [hunk 1]
```diff
@@ -11,47 +11,50 @@ import type {
...
+  InternalBundle,
+  InternalBundleGraph,
+  ParcelOptions,
+) => TBundle;
+
+export default class BundleGraph<TBundle: IBundle>
+  implements IBundleGraph<TBundle> {
   #graph; // InternalBundleGraph
   #options; // ParcelOptions
+  _createBundle: BundleFactory<TBundle>;
```
Comment (wbinnssmith) [inline packages/core/core/src/public/BundleGraph.js:46]: This doesn't seem to work as a private property (`#createBundle`) with Flow. The constructor is still accessible through the `constructor` property of the wrapper objects externally anyway.

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.36s | +56.00ms |
| Cached | 2.21s | -117.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 229.00ms | +57.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 230.00ms | +57.00ms ⚠️ |
| dist/legacy/kitchen-sink.679578a5.js | 1.16kb | +0.00b | 1.57s | +514.00ms ⚠️ |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 1.57s | +1.00s ⚠️ |
| dist/legacy/index.html | 709.00b | +0.00b | 587.00ms | -350.00ms 🚀 |
| dist/modern/index.html | 709.00b | +0.00b | 585.00ms | -349.00ms 🚀 |
| dist/legacy/styles.afb8e31a.css | 78.00b | +0.00b | 1.57s | -87.00ms 🚀 |
| dist/modern/styles.cda8c91b.css | 78.00b | +0.00b | 1.33s | -324.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 241.00ms | +32.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 241.00ms | +33.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.42s | +335.00ms |
| Cached | 2.27s | -16.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 226.00ms | -77.00ms 🚀 |
| dist/module/index.js | 37.47kb | +0.00b | 228.00ms | -75.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.67m | -4.45s |
| Cached | 4.97s | +26.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/ak-editor.1c8714e7.js | 2.50mb | -165.00b 🚀 | 1.51m | -3.35s |
| dist/EmojiPickerComponent.5fa86c7c.js | 98.96kb | -63.00b 🚀 | 15.58s | -245.00ms |
| dist/media-viewer.1fccd292.js | 71.13kb | +0.00b | 18.86s | -1.58s 🚀 |
| dist/card.47749cb1.js | 66.70kb | +0.00b | 18.85s | -1.59s 🚀 |
| dist/DatePicker.b3cbb2c4.js | 21.54kb | +0.00b | 8.11s | -3.70s 🚀 |
| dist/smartMediaEditor.ab4310d1.js | 14.62kb | +0.00b | 1.03m | -32.45s 🚀 |
| dist/EmojiTypeAheadComponent.899f7d25.js | 9.20kb | +0.00b | 48.44s | +31.07s ⚠️ |
| dist/dropzone.79e05ece.js | 3.44kb | +0.00b | 58.72s | -3.19s 🚀 |
| dist/ak-editor.47799eae.css | 3.43kb | +0.00b | 1.45m | +11.46s ⚠️ |
| dist/card.c002a0cc.js | 2.19kb | +0.00b | 53.20s | +21.26s ⚠️ |
| dist/media-picker-analytics-error-boundary.187d26a0.js | 1.82kb | +0.00b | 1.51m | +28.83s ⚠️ |
| dist/date.56d1aa9e.js | 1.81kb | +0.00b | 6.79s | -1.73s 🚀 |
| dist/ak-editor.7db5aa7b.js | 1.60kb | +0.00b | 1.24m | +10.09s ⚠️ |
| dist/16.6b0e338c.js | 1.38kb | +0.00b | 6.70s | -1.30s 🚀 |
| dist/16.81746ce1.js | 1.29kb | +0.00b | 6.70s | -1.30s 🚀 |
| dist/Modal.6509cbe3.js | 1.26kb | +0.00b | 6.92s | +393.00ms ⚠️ |
| dist/16.7ca41ffe.js | 1.16kb | +0.00b | 6.70s | -1.30s 🚀 |
| dist/mention.ae379620.js | 1.15kb | +0.00b | 8.51s | +1.48s ⚠️ |
| dist/quote.4733acb8.js | 1.10kb | +0.00b | 8.59s | +1.46s ⚠️ |
| dist/list.cf500a90.js | 1.03kb | +0.00b | 8.51s | +1.48s ⚠️ |
| dist/panel.e4fe05d1.js | 978.00b | +0.00b | 8.59s | +1.46s ⚠️ |
| dist/simpleHasher.dd9d4306.js | 609.00b | +0.00b | 15.51s | -34.00s 🚀 |
| dist/index.html | 219.00b | +0.00b | 938.00ms | -186.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/ak-editor.1c8714e7.js | 2.50mb | -161.00b 🚀 | 537.00ms | -61.00ms 🚀 |
| dist/pdfRenderer.74642224.js | 1.11mb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/editorView.54e7eb37.js | 586.92kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/popup.23ad5a3d.js | 218.72kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/ak-editor.f23d9d0b.js | 136.43kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/Toolbar.ff5ef36d.js | 100.36kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/EmojiPickerComponent.5fa86c7c.js | 98.96kb | -63.00b 🚀 | 535.00ms | -60.00ms 🚀 |
| dist/media-viewer.1fccd292.js | 71.13kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/card.47749cb1.js | 66.70kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/component.7a54e3cc.js | 55.63kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/Modal.dfa97108.js | 41.59kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/component.4c8e3dd6.js | 29.13kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/DatePicker.b3cbb2c4.js | 21.54kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/EmojiPickerComponent.a2474c55.js | 18.34kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/clipboard.8899f647.js | 17.18kb | +0.00b | 534.00ms | -62.00ms 🚀 |
| dist/js.baeda348.js | 16.46kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/dropzone.68826d45.js | 14.69kb | +0.00b | 534.00ms | -62.00ms 🚀 |
| dist/smartMediaEditor.ab4310d1.js | 14.62kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/ui.9e4d74b0.js | 14.45kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/workerHasher.ebf0b386.js | 11.83kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/EmojiTypeAheadComponent.899f7d25.js | 9.20kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/media-viewer.52d4959e.js | 4.18kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/EmojiPickerComponent.3f89a6ee.js | 3.92kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/16.cb2f1a3a.js | 3.72kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/Modal.1111ef83.js | 3.50kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/png-chunks-extract.c0f57e67.js | 3.49kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/dropzone.79e05ece.js | 3.44kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/ak-editor.47799eae.css | 3.43kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/16.c7be7376.js | 2.54kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/EmojiUploadComponent.4946a1c6.js | 2.50kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/ResourcedEmojiComponent.0979b683.js | 2.24kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/card.c002a0cc.js | 2.19kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/16.bb063a1f.js | 1.83kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.187d26a0.js | 1.82kb | +0.00b | 534.00ms | -62.00ms 🚀 |
| dist/date.56d1aa9e.js | 1.81kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/browser.90c1ebeb.js | 1.79kb | +0.00b | 534.00ms | -62.00ms 🚀 |
| dist/16.b1b0372d.js | 1.77kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/images.16b3af40.js | 1.75kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/feedback.b59078b6.js | 1.72kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/workerHasher.9975765a.js | 1.65kb | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/ak-editor.7db5aa7b.js | 1.60kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/list-number.d2097959.js | 1.54kb | +0.00b | 536.00ms | -61.00ms 🚀 |
| dist/status.f61d9ac5.js | 1.54kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.450f3a0d.js | 1.49kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/code.4289d7dd.js | 1.46kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/16.2b801dd1.js | 1.44kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/16.fe51c023.js | 1.44kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/16.7d4b2402.js | 1.42kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/link.36b44af7.js | 1.38kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/heading6.675c185b.js | 1.38kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.6b0e338c.js | 1.38kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/heading3.4e88bcd6.js | 1.37kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.e32aa49c.js | 1.34kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/16.50f9c790.js | 1.33kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/emoji.3654aec7.js | 1.31kb | +0.00b | 536.00ms | -61.00ms 🚀 |
| dist/media-card-analytics-error-boundary.9ca79f13.js | 1.31kb | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/16.81746ce1.js | 1.29kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/16.d1b58d7c.js | 1.29kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/16.1dbb5df4.js | 1.28kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/Modal.6509cbe3.js | 1.26kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/heading5.e338f1e8.js | 1.25kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.ceac57cd.js | 1.25kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/expand.4d883914.js | 1.24kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.0efadc10.js | 1.23kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/16.35be54a4.js | 1.23kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/16.6be9bbc8.js | 1.23kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/16.9015e8cf.js | 1.23kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/16.234ee133.js | 1.23kb | +0.00b | 537.00ms | -61.00ms 🚀 |
| dist/heading2.89a8e126.js | 1.19kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/16.7ca41ffe.js | 1.16kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/mention.ae379620.js | 1.15kb | +0.00b | 536.00ms | -60.00ms 🚀 |
| dist/heading4.b14d5ea4.js | 1.14kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/layout.d9df99ef.js | 1.13kb | +0.00b | 536.00ms | -62.00ms 🚀 |
| dist/divider.4476c3de.js | 1.11kb | +0.00b | 536.00ms | -61.00ms 🚀 |
| dist/quote.4733acb8.js | 1.10kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/action.b9409056.js | 1.08kb | +0.00b | 537.00ms | -60.00ms 🚀 |
| dist/decision.77ee0aaa.js | 1.07kb | +0.00b | 536.00ms | -61.00ms 🚀 |
| dist/panel-warning.e55309d2.js | 1.06kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/list.cf500a90.js | 1.03kb | +0.00b | 536.00ms | -61.00ms 🚀 |
| dist/heading1.165705a7.js | 1.03kb | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/media-card-analytics-error-boundary.a014425d.js | 995.00b | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/panel-error.364fbf3c.js | 987.00b | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/panel.e4fe05d1.js | 978.00b | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/table.fd1280b6.js | 970.00b | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/panel-success.66723ab0.js | 926.00b | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/panel-note.a8b696c4.js | 922.00b | +0.00b | 535.00ms | -61.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.ee954f0e.js | 892.00b | +0.00b | 534.00ms | -61.00ms 🚀 |
| dist/EmojiTypeAheadComponent.441c229c.js | 619.00b | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/simpleHasher.dd9d4306.js | 609.00b | +0.00b | 535.00ms | -60.00ms 🚀 |
| dist/index.html | 219.00b | +0.00b | 539.00ms | -60.00ms 🚀 |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/265077444342448641)**

File: packages/core/core/src/public/BundleGraph.js (hunk 1 above)
Comment (devongovett) [reply packages/core/core/src/public/BundleGraph.js:46]: Why doesn't it work?

Comment (mischnic) [reply packages/core/core/src/public/BundleGraph.js:46]: A bug in flow, private properties are not callable (though assigning them beforehand works): https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBjGBDAzrsAcTAG9UwwBiKOOAbnLEYzgDtcAXAJwFcMO4XABQBKMhQocAFgEtcAOmq0wAXjCjVAPjDceAUwYUAvugoAjLnADWe1hvESuejjy6sdshUrijDYE4wIglbYljysACb2jBQwzmAAHqoecoo09DFgTi5uib6MJkZAA

**Pull request merged**